### ISSUE TITLE
Add runtime exception handling

### ISSUE THREAD

**Issue by earncef** [d5/tengo#89]

**Comment by earncef** [comment]
@d5 

I've been working on adding exception catching feature. What do you think of something like this:

```
b := try {
    a := 2
    a()
}
out = b.value // out == "calling non-function: int"
```

**Comment by d5** [comment]
Thanks for the proposal! 

I admit Tengo's error handling is weak. I've been thinking about this too, but, I couldn't find a cleaner way to do this mostly because of the performance concerns. I'm not saying that error handling is less important than the running performance, but, I just want a balanced approach. We also need to think about adding more info on the reported error objects (like line number or even stack trace). It's important especially because the Tengo's more dynamic language, and, we will see more run-time errors than the compile-time errors for that reason. But again, it's about addressing all of them without losing too much on performance. 

Here are some of my initial thoughts on your proposal: 
- Tengo strictly differentiates statements vs expressions (like Go), especially the block statements (function, if/for statements) do not evaluate to a value (because they're not expressions). _I thought about the concept of "everything-is-expression" (like [Ponylang](https://tutorial.ponylang.io/expressions/control-structures.html)), but, it was way beyond the initial scope of this project._ Anyway, I think the block statements should not evaluates to a value to keep it consistent with everything else.
- I was hesitant to add explicit `try-catch` (and potentially `finally` clause too) because of its inevitable runtime overhead. I didn't experiment/validate it, but, my assumption is Go's `panic-recover` approach might be better (in terms of syntax and runtime performance).

**Comment by earncef** [comment]
I understand your consideration with block statements. 

I did try the full `try-catch` and it performs slower in the benchmarks vs a try{} block that evaluates to an error. Also, I'm not entirely sure catching an exception even makes much sense without differentiating between exception types and (possibly) changing the Error object's equality check.

The `panic-recover` approach would indeed be simpler. But I'm worried about the possible runtime overhead of a `defer`. I'll implement the `defer` and see how it performs.

**Comment by earncef** [comment]
Now that I'm working on implementing `defer` I've come across a bigger problem that might not solve the issue of panic-recover at all.

Tengo doesn't have named return values. This makes adding a defer to recover from a panic effectively of very little use as an error can't be passed along with the function's returning values.

Tengo also doesn't support multiple return values, so the go philosophy of returning values and errors and then checking for errors to ensure the validity of the return value also doesn't work.

**Comment by d5** [comment]
True. I've been working on some use cases _([FSM](https://github.com/d5/go-fsm) and couple other undisclosed projects)_, and, that way I'm learning a lot about the language/script needs, including this exception handling. Tengo definitely needs more work on error/exception handling/support, I agree. But it involves lots of bigger changes in the compiler/VM and also the language spec. _For that reasons, we may need to push this change to v2._ In the mean time, I am experimenting different approaches, but, I still don't have a good proposal that's balanced between simplicity and performance.

**Comment by d5** [comment]
@earncef: [some initial thoughts](https://github.com/d5/tengo/wiki/Error-Handling-V2) if you're interested. It's not finalized and there are lots of things missing there.

**Comment by earncef** [comment]
Some comments on the draft:

- Will the handles be called LIFO or FIFO?
- It's not very clear from your draft but I assume any handles declared after the statement that causes the error aren't executed?
- Handling some of the common runtime errors in a dynamically typed language like tengo would be tedious and clunky, which wouldn't be the case in Go as they'd end up as compile time errors. Consider the following. If `a` and `b` were non-indexable and cause a runtime error, it would be quite clunky to set them to a default value (say 0) and recover from it. 

```
myfunc := func(a, b) {
    handle err {
        d := b["abc"] // What if this is also an error? Do we handle inside a handle?
        return d
    }
    c := a["abc"]

    handle err {
        return c
    }
    d := b["abc"]

    return c+d
}

myfunc(indexableValue, anotherIndexableValue) //returns c+d
myfunc(indexableValue, nonIndexableValue) //returns c
myfunc(nonIndexableValue, indexableValue) //returns d
myfunc(nonIndexableValue, nonIndexableValue) //How to handle this? I suppose it could be handled on the function call, but I'm not sure that's realistic in more complex scenarios.
```

**Comment by d5** [comment]
Not directly related to this issue, but, this PR #103 adds more info (filename, line, column) to the runtime errors with little performance overhead. /cc @earncef
